crates/zk-core/src/check.rs

```diff
@@ -96,28 +100,53 @@ impl Check {
 
         let chi = Block::try_from(&transcript.finalize().as_bytes()[..16])
             .expect("block should be 16 bytes");
-        let chis = self.compute_chis(chi);
         let macs = mem::take(&mut self.triples);
+
+        // Computation with pre-split lanes.
+        const PARALLELISM: usize = 16;
+        let n = macs.len();
+        let segment_size = n.div_ceil(PARALLELISM);
+        let starts = Self::compute_chi_starts(chi);
+
+        let process_segment = |segment: &[Triple], mut rng: ChaCha12Rng| {
+            use rand_chacha::rand_core::RngCore;
+
+            let mut u_acc = Block::ZERO;
+            let mut v_acc = Block::ZERO;
+
+            for &triple in segment {
+                let mut chi_bytes = [0u8; 16];
+                rng.fill_bytes(&mut chi_bytes);
+                let chi = Block::from(chi_bytes);
+
+                let (u, v) = compute_terms(triple, chi);
+                u_acc ^= u;
+                v_acc ^= v;
+            }
+
+            (u_acc, v_acc)
+        };
+
         cfg_if! {
             if #[cfg(feature = "rayon")] {
                 use rayon::prelude::*;
 
                 let (mut u, mut v) = macs
-                    .into_par_iter()
-                    .zip(chis)
-                    .map(|(macs, chi)| compute_terms(macs, chi))
+                    .par_chunks(segment_size)
+                    .zip(starts.into_par_iter())
+                    .map(|(segment, chi_start)| process_segment(segment, chi_start))
```

Comment:
`map_with` will clone the PRG as needed. We ensure the PRG stream is always independent (and deterministic) by keying it to the chunk ID.

```suggestion
                    .enumerate()
                    .map_with(
                        rng,
                        |rng, (stream_id, segment)| {
                            rng.set_stream(stream_id as u64);
                            process_segment(rng, segment)
                        }
                    )
```